File: native-views/react-native-perp-depth-bar/android/src/main/java/com/margelo/nitro/perpdepthbar/HybridPerpSideRatio.kt
package com.margelo.nitro.perpdepthbar

import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.RectF
import android.view.Choreographer
import android.view.View
import com.facebook.proguard.annotations.DoNotStrip
import com.facebook.react.uimanager.ThemedReactContext
import kotlin.math.abs
import kotlin.math.exp

/**
 * Two horizontal segments whose widths ease to the bid/ask ratio.
 *
 * Animation model mirrors the depth bars: a [Choreographer] frame loop
 * continuously eases the split fraction toward its latest target (short fixed
 * exponential smoothing). New data just retargets; the loop stops once settled.
 */
@DoNotStrip
class HybridPerpSideRatio(val context: ThemedReactContext) : HybridPerpSideRatioSpec() {

  private val density = context.resources.displayMetrics.density
  private val bidPaint = Paint(Paint.ANTI_ALIAS_FLAG)
  private val askPaint = Paint(Paint.ANTI_ALIAS_FLAG)

  // Animated split fraction (bid share of the available width), 0..1.
  private var currentSplit = 0.5
  private var targetSplit = 0.5
  private var hasDrawn = false
  private var isDisposed = false
  // Set once the caller drives the ratio through `setRatio`. The
  // `bidPercentage`/`askPercentage` props are then ignored in `afterUpdate`,
  // otherwise every prop commit would reset the split back to the (static)
  // initial prop values (REACT-NATIVE-1JZ).
  private var imperativeRatio = false

  private val choreographer = Choreographer.getInstance()
  private var frameScheduled = false
  private var lastFrameNanos = 0L
  private val frameCallback = Choreographer.FrameCallback { onFrame(it) }

  override val view: View = object : View(context) {
    override fun onDraw(canvas: Canvas) {
      drawSegments(canvas)
    }
  }.apply {
    isClickable = false
    isFocusable = false
  }

  override var bidPercentage: Double = 50.0
    get() = field
    set(value) { field = value }

  override var askPercentage: Double = 50.0
    get() = field
    set(value) { field = value }

  override var segmentHeight: Double = 4.0
    get() = field
    set(value) { if (!isDisposed) { field = value; view.invalidate() } }

  override var cornerRadius: Double = 999.0
    get() = field
    set(value) { if (!isDisposed) { field = value; view.invalidate() } }

  override var gap: Double = 2.0
    get() = field
    set(value) { if (!isDisposed) { field = value; view.invalidate() } }

  /** Kept for OS "reduce motion" accessibility only — NOT a caller on/off knob. */
  override var reducedMotion: Boolean = false
    get() = field
    set(value) { field = value }

  override var longColor: String = ""
    get() = field
    set(value) {
      if (isDisposed) return
      field = value
      bidPaint.color = PerpColorParser.parse(value)
      view.invalidate()
    }

  override var shortColor: String = ""
    get() = field
    set(value) {
      if (isDisposed) return
      field = value
      askPaint.color = PerpColorParser.parse(value)
      view.invalidate()
    }

  /**
   * Imperative ratio update — bypasses the high-frequency
   * `bidPercentage`/`askPercentage` prop write path (props/JSICache lock
   * contention, REACT-NATIVE-1JZ). Companion to [HybridPerpDepthBars.setDepth].
   * The two args carry the SAME 0..100 semantics as the props of the same name;
   * we store them exactly like the prop setters and then route through the
   * identical [afterUpdate] clamp/split/frame-loop logic the prop path uses —
   * only the data source differs.
   */
  override fun setRatio(bidPercentage: Double, askPercentage: Double) {
    if (isDisposed) return
    imperativeRatio = true
    // Apply the args directly — do NOT route through the `bidPercentage`/
    // `askPercentage` prop fields, which a later prop commit could overwrite with
    // the static initial values.
    applyRatio(bidPercentage, askPercentage)
  }

  override fun afterUpdate() {
    super.afterUpdate()
    if (isDisposed) return
    // In imperative mode the ratio comes from `setRatio`; ignore the (static)
    // props so a prop commit can't reset the bar to the initial values.
    if (imperativeRatio) return
    applyRatio(bidPercentage, askPercentage)
  }

  private fun applyRatio(bidPercentage: Double, askPercentage: Double) {
    val bid = maxOf(bidPercentage, 1.0)
    val ask = maxOf(askPercentage, 1.0)
    val split = bid / (bid + ask)
    val snap = !hasDrawn || reducedMotion
    if (snap) {
      cancelFrameLoop()
      currentSplit = split
      targetSplit = split
      hasDrawn = true
      view.invalidate()
      return
    }
    targetSplit = split
    scheduleFrameLoop()
  }

  // MARK: - Continuous easing (Choreographer)
  private fun scheduleFrameLoop() {
    if (frameScheduled || isDisposed) return
    if (abs(targetSplit - currentSplit) <= 0.0005) { view.invalidate(); return }
    frameScheduled = true
    lastFrameNanos = 0L
    choreographer.postFrameCallback(frameCallback)
  }

  private fun cancelFrameLoop() {
    if (frameScheduled) {
      choreographer.removeFrameCallback(frameCallback)
      frameScheduled = false
    }
    lastFrameNanos = 0L
  }

  private fun onFrame(now: Long) {
    if (isDisposed) { frameScheduled = false; return }
    val dt = if (lastFrameNanos > 0L) (now - lastFrameNanos) / 1e9 else 1.0 / 60.0
    lastFrameNanos = now
    val alpha = 1.0 - exp(-maxOf(dt, 0.0) / TAU_SECONDS)

    val d = targetSplit - currentSplit
    if (abs(d) <= 0.0005) {
      currentSplit = targetSplit
      view.invalidate()
      frameScheduled = false
      lastFrameNanos = 0L
      return
    }
    currentSplit += d * alpha
    view.invalidate()
    choreographer.postFrameCallback(frameCallback)
  }

  private fun drawSegments(canvas: Canvas) {
    val w = view.width.toFloat()
    if (w <= 0f) return
    val h = (segmentHeight * density).toFloat()
    val g = (gap * density).toFloat()
    val y = (view.height - h) / 2f
    val available = (w - g).coerceAtLeast(0f)
    val bidW = available * currentSplit.toFloat()
    val askW = available - bidW
    val r = (cornerRadius * density).toFloat().coerceAtMost(h / 2f)

    if (bidW > 0f) {
      canvas.drawRoundRect(RectF(0f, y, bidW, y + h), r, r, bidPaint)
    }
    if (askW > 0f) {
      val askLeft = bidW + g
      canvas.drawRoundRect(RectF(askLeft, y, askLeft + askW, y + h), r, r, askPaint)
    }
  }

  override fun dispose() {
    isDisposed = true
    cancelFrameLoop()
  }

  companion object {
    /**
     * Exponential-smoothing time constant (seconds); mirror of
     * PerpTiming.sideRatioSmoothingTauSeconds on iOS.
     */
    private const val TAU_SECONDS = 0.12
  }
}

File: native-views/react-native-perp-depth-bar/ios/HybridPerpSideRatio.swift
import Foundation
import UIKit

/// Two horizontal segments whose widths ease to the bid/ask ratio.
///
/// Animation model mirrors the depth bars: a `CADisplayLink` continuously eases
/// the split fraction toward its latest target (short fixed exponential
/// smoothing). New data just retargets; the link stops once settled.
final class HybridPerpSideRatio: HybridPerpSideRatioSpec {

  // MARK: - HybridView
  var view: UIView = PerpLayoutView()

  private let bidLayer = CALayer()
  private let askLayer = CALayer()
  private var hasLaidOut = false

  // MARK: - Continuous-easing state (bid share of the available width, 0...1)
  private var currentSplit: CGFloat = 0.5
  private var targetSplit: CGFloat = 0.5
  private var displayLink: CADisplayLink?
  private var lastTick: CFTimeInterval = 0

  // MARK: - Props
  var bidPercentage: Double = 50 { didSet { scheduleLayout() } }
  var askPercentage: Double = 50 { didSet { scheduleLayout() } }
  var segmentHeight: Double = 4 { didSet { scheduleLayout() } }
  var cornerRadius: Double = 999 { didSet { scheduleLayout() } }
  var gap: Double = 2 { didSet { scheduleLayout() } }
  /// Kept for OS "reduce motion" accessibility only — NOT a caller on/off knob.
  var reducedMotion: Bool = false { didSet { scheduleLayout() } }

  var longColor: String = "" {
    didSet {
      bidLayer.backgroundColor = PerpColorParser.parse(longColor).cgColor
    }
  }
  var shortColor: String = "" {
    didSet {
      askLayer.backgroundColor = PerpColorParser.parse(shortColor).cgColor
    }
  }

  override init() {
    super.init()
    view.isUserInteractionEnabled = false
    view.clipsToBounds = true
    bidLayer.maskedCorners = [.layerMinXMinYCorner, .layerMinXMaxYCorner]
    askLayer.maskedCorners = [.layerMaxXMinYCorner, .layerMaxXMaxYCorner]
    view.layer.addSublayer(bidLayer)
    view.layer.addSublayer(askLayer)
    if let layoutView = view as? PerpLayoutView {
      layoutView.onLayout = { [weak self] in self?.performLayout() }
    }
  }

  deinit {
    stopDisplayLink()
  }

  /// Imperative ratio update — bypasses the high-frequency
  /// `bidPercentage`/`askPercentage` prop write path (props/JSICache lock
  /// contention, REACT-NATIVE-1JZ). Companion to `HybridPerpDepthBars.setDepth`.
  /// The two args carry the SAME 0..100 semantics as the props of the same
  /// name; assigning them feeds through the identical `didSet` →
  /// `scheduleLayout` → `performLayout` clamp/split/display-link path the prop
  /// setters use — only the data source differs.
  func setRatio(bidPercentage: Double, askPercentage: Double) throws {
    self.bidPercentage = bidPercentage
    self.askPercentage = askPercentage
  }

  private func scheduleLayout() {
    // Imperative setRatio runs on the JS thread (off-main); `setNeedsLayout` must
    // be issued on the main thread.
    if Thread.isMainThread {
      view.setNeedsLayout()
    } else {
      DispatchQueue.main.async { [weak self] in self?.view.setNeedsLayout() }
    }
  }

  private func performLayout() {
    let bounds = view.bounds
    guard bounds.width > 0 else { return }

    let h = CGFloat(segmentHeight)
    CATransaction.begin()
    CATransaction.setDisableActions(true)
    bidLayer.cornerRadius = min(CGFloat(cornerRadius), h / 2)
    askLayer.cornerRadius = min(CGFloat(cornerRadius), h / 2)
    CATransaction.commit()

    let bid = max(bidPercentage, 1)
    let ask = max(askPercentage, 1)
    let target = CGFloat(bid / (bid + ask))

    let snap = !hasLaidOut || reducedMotion
    if snap {
      stopDisplayLink()
      currentSplit = target
      targetSplit = target
      layoutSegments(currentSplit)
    } else {
      targetSplit = target
      layoutSegments(currentSplit) // apply current + any geometry change now
      startDisplayLinkIfNeeded()
    }
    hasLaidOut = true
  }

  /// Lays both segment frames out from a split fraction (no implicit anim).
  private func layoutSegments(_ split: CGFloat) {
    let bounds = view.bounds
    guard bounds.width > 0 else { return }
    let h = CGFloat(segmentHeight)
    let y = (bounds.height - h) / 2
    let g = CGFloat(gap)
    let available = max(bounds.width - g, 0)
    let bidW = available * split
    let askW = max(available - bidW, 0)
    CATransaction.begin()
    CATransaction.setDisableActions(true)
    bidLayer.frame = CGRect(x: 0, y: y, width: bidW, height: h)
    askLayer.frame = CGRect(x: bidW + g, y: y, width: askW, height: h)
    CATransaction.commit()
  }

  // MARK: - Continuous easing (display link)
  private func startDisplayLinkIfNeeded() {
    if displayLink != nil { return }
    guard abs(targetSplit - currentSplit) > 0.0005 else { return }
    lastTick = 0
    let proxy = SideRatioLinkProxy(self)
    let link = CADisplayLink(target: proxy, selector: #selector(SideRatioLinkProxy.tick(_:)))
    link.add(to: .main, forMode: .common)
    displayLink = link
  }

  private func stopDisplayLink() {
    displayLink?.invalidate()
    displayLink = nil
    lastTick = 0
  }

  func handleTick(_ link: CADisplayLink) {
    let now = link.timestamp
    let dt = lastTick > 0 ? now - lastTick : link.duration
    lastTick = now
    let tau = PerpTiming.sideRatioSmoothingTauSeconds
    let alpha = CGFloat(1 - exp(-max(dt, 0) / max(tau, 0.0001)))

    let d = targetSplit - currentSplit
    if abs(d) <= 0.0005 {
      if currentSplit != targetSplit {
        currentSplit = targetSplit
        layoutSegments(currentSplit)
      }
      stopDisplayLink()
      return
    }
    currentSplit += d * alpha
    layoutSegments(currentSplit)
  }
}

/// Weak forwarder so the `CADisplayLink` does not retain `HybridPerpSideRatio`.
private final class SideRatioLinkProxy {
  weak var owner: HybridPerpSideRatio?
  init(_ owner: HybridPerpSideRatio) { self.owner = owner }
  @objc func tick(_ link: CADisplayLink) { owner?.handleTick(link) }
}
